### PR TITLE
Handler identity claim update for JIT provisioned users

### PR DESCRIPTION
**Pull request by GANGANI**
### Description

This PR is to handle identity claim update for JIT provisioned users as all attribute update is blocked for JIT provisioned users.

Related issue:https://github.com/wso2-enterprise/asgardeo-product/issues/5315

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 1]
```diff
@@ -169,9 +172,11 @@ public boolean doPostSetUserClaimValueWithID(String s, UserStoreManager userStor
     public boolean doPreSetUserClaimValueWithID(String userID, String claimURI, String claimValue, String profileName,
                                                 UserStoreManager userStoreManager) throws UserStoreException {
 
-        // Validate whether claim update request is for a provisioned user.
-        if (FrameworkUtils.isJITProvisionEnhancedFeatureEnabled()) {
-            validateClaimUpdate(getUsernameFromUserID(userID, userStoreManager));
+        if (FrameworkUtils.isJITProvisionEnhancedFeatureEnabled() && StringUtils.isNotBlank(claimURI)) {
+            // Check whether if this is an account locking scenario.
+            boolean isAccountLockRequest = claimURI.equals(ACCOUNT_LOCK_CLAIM_URI);
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:177]: Any possibility of NPE? If so better to use StringUtils.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 2]
```diff
@@ -245,19 +254,25 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
...
 
     public boolean doPreSetUserClaimValue(String userName, String claimURI, String claimValue, String profileName,
                                           UserStoreManager userStoreManager) throws UserStoreException {
 
-        // Validate whether claim update request is for a provisioned user.
-        if (FrameworkUtils.isJITProvisionEnhancedFeatureEnabled()) {
-            validateClaimUpdate(userName);
+        if (FrameworkUtils.isJITProvisionEnhancedFeatureEnabled() && StringUtils.isNotBlank(claimURI)) {
+            // Check whether if this is an account locking scenario.
+            boolean isAccountLockRequest = claimURI.equals(ACCOUNT_LOCK_CLAIM_URI);
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:273]: Same as above comment.

Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/376/commits/01ac2420b372fb1cec7701640adfb172f6681823

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 1 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/376/commits/01ac2420b372fb1cec7701640adfb172f6681823

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 3]
```diff
@@ -169,8 +169,8 @@ public boolean doPostSetUserClaimValueWithID(String s, UserStoreManager userStor
     public boolean doPreSetUserClaimValueWithID(String userID, String claimURI, String claimValue, String profileName,
                                                 UserStoreManager userStoreManager) throws UserStoreException {
 
-        // Validate whether claim update request is for a provisioned user.
-        if (FrameworkUtils.isJITProvisionEnhancedFeatureEnabled()) {
+        if (FrameworkUtils.isJITProvisionEnhancedFeatureEnabled() && StringUtils.isNotBlank(claimURI) &&
+                !claimURI.contains(UserCoreConstants.ClaimTypeURIs.IDENTITY_CLAIM_URI)) {
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:173]: Can't we use `isIdentityClaimUpdate` here as well?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 4]
```diff
@@ -539,4 +542,16 @@ public boolean doPostGetUserClaimValuesWithID(String userID, String[] claims, St
         claimMap.put(SCIMCommonConstants.READ_ONLY_USER_CLAIM, String.valueOf(userStoreManager.isReadOnly()));
         return true;
     }
+
+    private boolean isIdentityClaimUpdate(Map<String, String> claims) {
+
+        boolean isIdentityClaimUpdate = false;
+        for (Map.Entry<String, String> claim : claims.entrySet()) {
+            if (claim.getKey().contains(UserCoreConstants.ClaimTypeURIs.IDENTITY_CLAIM_URI)) {
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:550]: We need to check `contains` or `startsWith`?

Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/376/commits/885367b448d53ffdb5414c784b8fccf532677ebb

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 3 above)
Comment (GANGANI) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: Fixed https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/376/commits/885367b448d53ffdb5414c784b8fccf532677ebb

**Review by ashensw — APPROVED** [review]

**Pull request merged**